### PR TITLE
Allow conversion from `Report` to `Error`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

It's currently hard to deal with `Report` when an `Error` is required. This PR adds conversions from `Report` to `Error`.

## 🔗 Related links

- resolves https://github.com/hashintel/hash/issues/1746

## 🔍 What does this change?

- Add `Report::into_error` to convert `impl Error`
- Add `Report::as_error` to return `&impl Error`
- Add `From` implementation of `Report` for `Box<dyn Error>` (+ `Send` and/or `Sync`)

## 📜 Does this require a change to the docs?

The changelog was changed to reflect these changes

## 🛡 What tests cover this?

Three new tests were added:
- A UI test to make sure, that `into_report` cannot be called on `Result<T, Report<C>>`
- Two tests to check the new methods/trait implementation by converting a `Report` to `Error` and ensure, that the provided values are still accessible

### CODE REVIEW

File: packages/libs/error-stack/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+impl<C> fmt::Display for ReportError<C> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.0, fmt)
+    }
+}
+
+impl<C> Error for ReportError<C> {
+    #[cfg(nightly)]
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+        self.0.frames().for_each(|frame| frame.provide(demand));
```
Comment (indietyp) [inline packages/libs/error-stack/src/error.rs:41]: Out of curiosity, any reason for choosing `for_each` over `for frame in self.frames()`?

File: packages/libs/error-stack/src/lib.rs [hunk 2]
```diff
@@ -464,7 +464,6 @@
 )]
 
 extern crate alloc;
-extern crate core;
```
Comment (indietyp) [inline packages/libs/error-stack/src/lib.rs:467]: I love how CLion always automatically adds this line...

**Review by indietyp — APPROVED** [review]
Looks good to me, just some minor non-blocking comments. I already noted this over private DMs, but with this change

```rust
#[test]
fn as_error_nest() {
    let report = create_report();
    let base = report.into_error();
    let report = Report::new(base);

    println!("{report:?}");
}
```

is possible. I think that is ok, as it does not break formatting and has quite the high barrier for creating it.

File: packages/libs/error-stack/src/error.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/libs/error-stack/src/error.rs:41]: No specific reason. I used it here as it's only one line vs three lines.

**Pull request merged**